module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,454 @@
+
+// New creates a new execution data requester component
+func New(
+	log zerolog.Logger,
+	edrMetrics module.ExecutionDataRequesterMetrics,
+	datastore datastore.Batching,
+	blobservice network.BlobService,
+	eds state_synchronization.ExecutionDataService,
+	processedHeight storage.ConsumerProgress,
+	processedNotifications storage.ConsumerProgress,
+	state protocol.State,
+	headers storage.Headers,
+	results storage.ExecutionResults,
+	cfg ExecutionDataConfig,
+) (state_synchronization.ExecutionDataRequester, error) {
+	e := &executionDataRequester{
+		log:                  log.With().Str("component", "execution_data_requester").Logger(),
+		ds:                   datastore,
+		bs:                   blobservice,
+		eds:                  eds,
+		metrics:              edrMetrics,
+		headers:              headers,
+		results:              results,
+		config:               cfg,
+		finalizationNotifier: engine.NewNotifier(),
+	}
+
+	executionDataNotifier := engine.NewNotifier()
+
+	// jobqueue Jobs object that tracks sealed blocks by height. This is used by the blockConsumer
+	// to get a sequential list of sealed blocks.
+	sealedBlockReader := jobqueue.NewSealedBlockHeaderReader(state, headers)
+
+	// blockConsumer ensures every sealed block's execution data is downloaded.
+	// It listens to block finalization events from `finalizationNotifier`, then checks if there
+	// are new sealed blocks with `sealedBlockReader`. If there are, it starts workers to process
+	// them with `processingBlockJob`, which fetches execution data. At most `fetchWorkers` workers
+	// will be created for concurrent processing. When a sealed block's execution data has been
+	// downloaded, it updates and persists the highest consecutive downloaded height with
+	// `processedHeight`. That way, if the node crashes, it reads the `processedHeight` and resume
+	// from `processedHeight + 1`. If the database is empty, rootHeight will be used to init the
+	// last processed height. Once the execution data is fetched and stored, it notifies
+	// `executionDataNotifier`.
+	e.blockConsumer = jobqueue.NewComponentConsumer(
+		log.With().Str("module", "block_consumer").Logger(),
+		e.finalizationNotifier.Channel(), // to listen to finalization events to find newly sealed blocks
+		processedHeight,                  // read and persist the downloaded height
+		sealedBlockReader,                // read sealed blocks by height
+		e.config.InitialBlockHeight,      // initial "last processed" height for empty db
+		e.processBlockJob,                // process the sealed block job to download its execution data
+		fetchWorkers,                     // the number of concurrent workers
+		e.config.MaxSearchAhead,          // max number of unsent notifications to allow before pausing new fetches
+	)
+	// notifies notificationConsumer when new ExecutionData blobs are available
+	e.blockConsumer.SetPostNotifier(func(module.JobID) { executionDataNotifier.Notify() })
```

Comment:
using the `PreNotifier` here introduces a race condition between the `blockConsumer` updating its `lastProcessedIndex`, and the `notificationConsumer` checking the next available job via `executionDataReader`.

Since the job queues are pipelined, if `notificationConsumer` checks for new jobs first, it will not find any new jobs available. At the end of a spork, this could cause execution data to not be fetched. It should recover after a reboot.

I also think it's acceptable to require consumers of the execution data received notifications to track which blocks they've received and backfill any that were missed on restart. That was the original design, but the jobqueue made it easy to backfill automatically.

Reply:
Yeah, good point. You did right, we should use `SetPostNotifier`, not `SetPreNotifier`.

Maybe we should explain this a bit. 
```suggestion
	// SetPostNotifier will notify executionDataNotifier AFTER e.blockConsumer.LastProcessedIndex is updated.
	// Even though it doesn't guarantee to notify for every height at least once, the notificationConsumer is
	// able to guarantee to process every height at least once, because the notificationConsumer finds new job 
	// using executionDataReader which finds new height using e.blockConsumer.LastProcessedIndex
	e.blockConsumer.SetPostNotifier(func(module.JobID) { executionDataNotifier.Notify() })
```

Reply:
Some idea to share for future:

I feel building this "pipeline" with 2 job queues involves quite a few boilerplate code that requires to know some implementation detail of how the job queue works. Like the next job queue needs to use previous job's LastProcessedIndex to find new jobs, and the previous job queue needs to use SetPostNotifier to notify the next job queue. 

If this "2 job queue pipeline" scenario is common enough, we could consider implementing a "wiring" function that wires two job queues into a pipeline in the future.

Reply:
Yea, I agree.